File: assets/codebase/rendering/DemoScenePlatformInstructionIconSetComponent.cs
namespace city.rendering {
    /// <summary>
    /// Enables exactly one platform-specific instruction-icon group beneath the owning entity so demo-scene overlays show the best controller family per runtime platform.
    /// </summary>
    public sealed class DemoScenePlatformInstructionIconSetComponent : UpdateComponent {
        /// <summary>
        /// Stable direct-child index used for the Xbox 360 style instruction icon group.
        /// </summary>
        const int Xbox360GroupChildIndex = 0;

        /// <summary>
        /// Stable direct-child index used for the PS2 style instruction icon group.
        /// </summary>
        const int Ps2GroupChildIndex = 1;

        /// <summary>
        /// Stable direct-child index used for the Switch style instruction icon group.
        /// </summary>
        const int SwitchGroupChildIndex = 2;

        /// <summary>
        /// Caches the owning icon-set host entity once the component attaches to a scene hierarchy.
        /// </summary>
        Entity OwnerEntity;

        /// <summary>
        /// Tracks whether the platform-specific child-group enablement has already been applied successfully.
        /// </summary>
        bool IsConfigured;

        /// <summary>
        /// Captures the owning entity and attempts to apply the runtime platform icon selection immediately.
        /// </summary>
        /// <param name="entity">Owning entity that hosts the platform icon groups.</param>
        public override void ComponentAdded(Entity entity) {
            base.ComponentAdded(entity);
            if (entity == null) {
                throw new ArgumentNullException(nameof(entity));
            }

            OwnerEntity = entity;
            IsConfigured = false;
            TryApplyPlatformSelection();
        }

        /// <summary>
        /// Retries platform-group enablement until the overlay hierarchy is fully available.
        /// </summary>
        public override void Update() {
            if (IsConfigured) {
                return;
            }

            TryApplyPlatformSelection();
        }

        /// <summary>
        /// Clears cached runtime state when the component detaches from the scene hierarchy.
        /// </summary>
        /// <param name="entity">Owning entity that is removing this component.</param>
        public override void ComponentRemoved(Entity entity) {
            OwnerEntity = null;
            IsConfigured = false;
            base.ComponentRemoved(entity);
        }

        /// <summary>
        /// Resolves the best platform icon family for the current runtime and applies it to the direct child groups.
        /// </summary>
        void TryApplyPlatformSelection() {
            if (OwnerEntity == null || OwnerEntity.Children == null || OwnerEntity.Children.Count == 0) {
                return;
            } else if (Core.Instance == null || Core.Instance.PlatformInfo == null || string.IsNullOrWhiteSpace(Core.Instance.PlatformInfo.Name)) {
                return;
            }

            int selectedGroupIndex = ResolveSelectedGroupIndex(Core.Instance.PlatformInfo.Name);
            for (int childIndex = 0; childIndex < OwnerEntity.Children.Count; childIndex++) {
                Entity childEntity = OwnerEntity.Children[childIndex];
                if (childEntity == null) {
                    continue;
                }

                childEntity.Enabled = childIndex == selectedGroupIndex;
            }

            IsConfigured = true;
        }

        /// <summary>
        /// Maps one runtime platform name to the instruction icon family that best matches the platform's controller conventions.
        /// </summary>
        /// <param name="platformName">Runtime platform name exposed by the packaged player.</param>
        /// <returns>Stable direct-child index that should remain enabled.</returns>
        int ResolveSelectedGroupIndex(string platformName) {
            if (string.IsNullOrWhiteSpace(platformName)) {
                throw new ArgumentException("Platform name must be provided.", nameof(platformName));
            }

            string normalizedPlatformName = platformName.Trim().ToLowerInvariant();
            if (normalizedPlatformName.Contains("3ds", StringComparison.Ordinal) || normalizedPlatformName == "ds") {
                return SwitchGroupChildIndex;
            } else if (normalizedPlatformName.Contains("ps2", StringComparison.Ordinal) || normalizedPlatformName.Contains("psp", StringComparison.Ordinal)) {
                return Ps2GroupChildIndex;
            } else if (normalizedPlatformName.Contains("windows", StringComparison.Ordinal)
                || normalizedPlatformName.Contains("win32", StringComparison.Ordinal)
                || normalizedPlatformName.Contains("gamecube", StringComparison.Ordinal)) {
                return Xbox360GroupChildIndex;
            }

            return Xbox360GroupChildIndex;
        }
    }
}

File: windows-build-20260602-stack-box-direct-real-bepu-fix2/cooked/code/gameplay/DemoScenePlatformInstructionIconSetComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "DemoScenePlatformInstructionIconSetComponent.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_string.hpp"
#include "DemoScenePlatformInstructionIconSetComponent.hpp"
#include "runtime/native_list.hpp"
#include "Component.hpp"
#include "Core.hpp"
#include "Entity.hpp"
#include "PlatformInfo.hpp"
#include "UpdateComponent.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_list.hpp"
#include "runtime/native_string.hpp"

DemoScenePlatformInstructionIconSetComponent::DemoScenePlatformInstructionIconSetComponent() : IsConfigured(), OwnerEntity()
{
}

void DemoScenePlatformInstructionIconSetComponent::ComponentAdded(Entity* entity)
{
UpdateComponent::ComponentAdded(entity);
    if (entity == nullptr)
    {
throw new ArgumentNullException("entity");
    }
this->OwnerEntity = entity;
this->IsConfigured = false;
this->TryApplyPlatformSelection();
}

void DemoScenePlatformInstructionIconSetComponent::ComponentRemoved(Entity* entity)
{
this->OwnerEntity = nullptr;
this->IsConfigured = false;
UpdateComponent::ComponentRemoved(entity);
}

void DemoScenePlatformInstructionIconSetComponent::Update()
{
    if (this->IsConfigured)
    {
return;    }
this->TryApplyPlatformSelection();
}

bool DemoScenePlatformInstructionIconSetComponent::ContainsNormalizedPlatformToken(std::string normalizedPlatformName, std::string token)
{
    if (String::IsNullOrWhiteSpace(normalizedPlatformName))
    {
throw ([&]() {
auto __ctor_arg_00000092 = "Normalized platform name must be provided.";
auto __ctor_arg_00000093 = "normalizedPlatformName";
return new ArgumentException(__ctor_arg_00000092, __ctor_arg_00000093);
})();
    }
else {
    if (String::IsNullOrWhiteSpace(token))
    {
throw ([&]() {
auto __ctor_arg_00000094 = "Platform token must be provided.";
auto __ctor_arg_00000095 = "token";
return new ArgumentException(__ctor_arg_00000094, __ctor_arg_00000095);
})();
    }
else {
    if (static_cast<int32_t>(token.size()) > static_cast<int32_t>(normalizedPlatformName.size()))
    {
return false;    }
}
}
const int32_t lastStartIndex = static_cast<int32_t>(normalizedPlatformName.size()) - static_cast<int32_t>(token.size());
for (int32_t startIndex = 0; startIndex <= lastStartIndex; startIndex++) {
bool matched = true;
for (int32_t tokenIndex = 0; tokenIndex < static_cast<int32_t>(token.size()); tokenIndex++) {
    if (normalizedPlatformName[startIndex + tokenIndex] == token[tokenIndex])
    {
continue;
    }
matched = false;
break;
}
    if (matched)
    {
return true;    }
}
return false;}

int32_t DemoScenePlatformInstructionIconSetComponent::ResolveSelectedGroupIndex(std::string platformName)
{
    if (String::IsNullOrWhiteSpace(platformName))
    {
throw ([&]() {
auto __ctor_arg_00000096 = "Platform name must be provided.";
auto __ctor_arg_00000097 = "platformName";
return new ArgumentException(__ctor_arg_00000096, __ctor_arg_00000097);
})();
    }
const std::string normalizedPlatformName = String::ToLowerInvariant(String::Trim(platformName));
    if (this->ContainsNormalizedPlatformToken(normalizedPlatformName, "3ds") || normalizedPlatformName == "ds")
    {
return SwitchGroupChildIndex;    }
else {
    if (this->ContainsNormalizedPlatformToken(normalizedPlatformName, "ps2") || this->ContainsNormalizedPlatformToken(normalizedPlatformName, "psp"))
    {
return Ps2GroupChildIndex;    }
else {
    if (this->ContainsNormalizedPlatformToken(normalizedPlatformName, "windows") || this->ContainsNormalizedPlatformToken(normalizedPlatformName, "win32") || this->ContainsNormalizedPlatformToken(normalizedPlatformName, "gamecube"))
    {
return Xbox360GroupChildIndex;    }
}
}
return Xbox360GroupChildIndex;}

void DemoScenePlatformInstructionIconSetComponent::TryApplyPlatformSelection()
{
    if (this->OwnerEntity == nullptr || this->OwnerEntity->get_Children() == nullptr || this->OwnerEntity->get_Children()->get_Count() == 0)
    {
return;    }
else {
    if (Core::get_Instance() == nullptr || Core::get_Instance()->get_PlatformInfo() == nullptr || String::IsNullOrWhiteSpace(Core::get_Instance()->get_PlatformInfo()->get_Name()))
    {
return;    }
}
const int32_t selectedGroupIndex = this->ResolveSelectedGroupIndex(Core::get_Instance()->get_PlatformInfo()->get_Name());
for (int32_t childIndex = 0; childIndex < this->OwnerEntity->get_Children()->get_Count(); childIndex++) {
Entity *childEntity = (*this->OwnerEntity->get_Children()).get_Item(childIndex);
    if (childEntity == nullptr)
    {
continue;
    }
childEntity->set_Enabled(childIndex == selectedGroupIndex);
}
this->IsConfigured = true;
}

